packages/rum-core/src/domain/rumEventsCollection/view/trackInteractionToNextPaint.ts

```diff
@@ -0,0 +1,114 @@
+import { type Duration, noop, isExperimentalFeatureEnabled, ExperimentalFeature } from '@datadog/browser-core'
+import {
+  supportPerformanceTimingEvent,
+  type RumEventTiming,
+  type RumFirstInputTiming,
+} from '../../../browser/performanceCollection'
+import { LifeCycleEventType, type LifeCycle } from '../../lifeCycle'
+import { ViewLoadingType } from '../../../rawRumEvent.types'
+import { getInteractionCount, initInteractionCountPolyfill } from './interactionCountPolyfill'
+
+// Arbitrary value to prevent unnecessary memory usage on views with lots of interactions.
+const MAX_INTERACTION_ENTRIES = 10
+
+/**
+ * Track the interaction to next paint (INP).
+ * To avoid outliers, return the p98 worst interaction of the view.
+ * Documentation: https://web.dev/inp/
+ * Reference implementation: https://github.com/GoogleChrome/web-vitals/blob/main/src/onINP.ts
+ */
+export function trackInteractionToNextPaint(viewLoadingType: ViewLoadingType, lifeCycle: LifeCycle) {
+  if (
+    !isInteractionToNextPaintSupported() ||
+    !isExperimentalFeatureEnabled(ExperimentalFeature.INTERACTION_TO_NEXT_PAINT)
+  ) {
+    return {
+      getInteractionToNextPaint: () => undefined,
+      stop: noop,
+    }
+  }
+
+  // List of longest interactions on the view by duration.
+  const longestInteractions: Array<RumEventTiming | RumFirstInputTiming> = []
+
+  const { getViewInteractionCount } = trackViewInteractionCount(viewLoadingType)
+  let maxInpDuration = -1 as Duration
+
+  const { unsubscribe: stop } = lifeCycle.subscribe(LifeCycleEventType.PERFORMANCE_ENTRIES_COLLECTED, (entries) => {
+    for (const entry of entries) {
+      if (entry.entryType === 'event' && entry.interactionId) {
```

Comment:
Compared to Goole implementation: No need to track 'first-input' event because our duration threshold is at 40 and so first-input above that duration will also be reported as 'event' entry.